src/graph.cc

```diff
@@ -284,7 +283,7 @@ bool DependencyScan::RecomputeOutputDirty(const Edge* edge,
         EXPLAIN("command line changed for %s", output->path().c_str());
         return true;
       }
-      if (most_recent_input && entry->mtime < most_recent_input->mtime()) {
+      if (!used_log && most_recent_input && entry->mtime < most_recent_input->mtime()) {
```

Comment:
Because we've already performed this check on line 265 if `used_log` is true. I think this is actually dead code looking at it now, I can probably assert that `used_log == true` here and remove the whole if scope.

Reply:
Ah, now I remember. that check is only performed if the initial `output->mtime() < most_recent_input->mtime()` was true on line 255. If that's false, we still want to do this check here, in which case `used_log` would still be false.

Reply:
Would it be possible to add a unit test for this?

Reply:
There are already multiple tests that test this conditional:

`used_log == false`:

    1. RebuildAfterFailure
    2. RebuildWithNoInputs
    3. RestatTest
    4. and more
 
`used_log == true`:

    1. RestatTest
    2. RestatMissingDepfileDepslog

Reply:
I'll have a closer look, thx.

Reply:
What do you think about the following?

```cpp
bool DependencyScan::RecomputeOutputDirty(const Edge* edge,
                                          const string& command,
                                          Node* output) {
  if (edge->is_phony()) {
    // Phony edges don't write any output.  Outputs are only dirty if
    // there are no inputs and we're missing the output.
    if (edge->inputs_.empty() && !output->exists()) {
      EXPLAIN("output %s of phony edge with no inputs doesn't exist",
              output->path().c_str());
      return true;
    }
    return false;
  }

  BuildLog::LogEntry* entry = 0;

  // Dirty if we're missing the output.
  if (!output->exists()) {
    EXPLAIN("output %s doesn't exist", output->path().c_str());
    return true;
  }

  const Node* most_recent_input = edge->most_recent_input_;

  if (build_log()) {
    bool generator = edge->GetBindingBool("generator");
    if (entry || (entry = build_log()->LookupByOutput(output->path()))) {
      if (!generator &&
          BuildLog::LogEntry::HashCommand(command) != entry->command_hash) {
        // May also be dirty due to the command changing since the last build.
        // But if this is a generator rule, the command changing does not make us
        // dirty.
        EXPLAIN("command line changed for %s", output->path().c_str());
        return true;
      }
      if (most_recent_input && entry->mtime < most_recent_input->mtime()) {
        // May also be dirty due to the mtime in the log being older than the
        // mtime of the most recent input.  This can occur even when the mtime
        // on disk is newer if a previous run wrote to the output file but
        // exited with an error or was interrupted.
        EXPLAIN("recorded mtime of %s older than most recent input %s (%" PRId64 " vs %" PRId64 ")",
                output->path().c_str(), most_recent_input->path().c_str(),
                entry->mtime, most_recent_input->mtime());
        return true;
      }
    }
    if (!entry && !generator) {
      EXPLAIN("command line not found in log for %s", output->path().c_str());
      return true;
    }
  } else if (most_recent_input &&
             output->mtime() < most_recent_input->mtime()) {
    EXPLAIN(
        "output %s older than most recent input %s "
        "(%" PRId64 " vs %" PRId64 ")",
        output->path().c_str(), most_recent_input->path().c_str(),
        output->mtime(), most_recent_input->mtime());
    return true;
  }

  return false;
}
```

It's a bit shorter and avoids the `used_log` bool (which made it harder to understand for me what is going on).

Reply:
Works for me. I'll try it out and make sure the tests still pass. If so, I'll make this change.

Reply:
Done.